### PR TITLE
Add support for group slurm jobs and optimize packaging

### PR DESCRIPTION
**Pull request by hemildesai**

### CODE REVIEW

**Review by Kipok — APPROVED** [review]

**Comment by Kipok** [comment]
After some more testing I realized that we actually need to have that code block that monitors if any of the tasks finishes and kills everything if that's the case. @hemildesai is it possible to enable it or maybe have a parameter that controls whether it should be enabled? The use-case is that we have a server that's always running and then a client that talks to that server, so after client finishes, we need to kill the server to be able to free the node. That can be done either by waiting just for the client or by stopping if at least one of the jobs is over

**Comment by hemildesai** [comment]
> After some more testing I realized that we actually need to have that code block that monitors if any of the tasks finishes and kills everything if that's the case. @hemildesai is it possible to enable it or maybe have a parameter that controls whether it should be enabled? The use-case is that we have a server that's always running and then a client that talks to that server, so after client finishes, we need to kill the server to be able to free the node. That can be done either by waiting just for the client or by stopping if at least one of the jobs is over

Added in https://github.com/NVIDIA/NeMo-Run/pull/24/commits/dce4ba56d1c50642922ff363ae55856ec4e967b7

File: src/nemo_run/core/execution/slurm.py [hunk 1]
```diff
@@ -879,26 +880,69 @@ def materialize(self) -> str:
...
+                else:
+                    _container_flags = get_container_flags(
+                        base_mounts=self.slurm_config.container_mounts,
+                        src_job_dir=os.path.join(
+                            slurm_job_dir,
+                            job_directory_name,
+                        ),
+                        container_image=self.slurm_config.container_image,
+                    )
+                    _srun_args = self.slurm_config.srun_args or []
```
Comment (Kipok) [inline src/nemo_run/core/execution/slurm.py:915]: Can you please move the user-provided arguments to the end of the list if possible? This way I can override the default waiting time. Or make it customizable somehow, as it becomes quite inconvenient to wait for so long if we are debugging something (and we do often debug things on slurm when developing new pipelines)

File: test/core/execution/artifacts/group_resource_req_slurm.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            # Process is still running.
+            all_done=false
+        else
+            # Process is no longer running => check its exit status.
+            wait "$pid"
+            exitcode=$?
+            echo "Process $pid exited with code $exit_code at $(date '+%Y-%m-%d %H:%M:%S')"
+            # Wait a bit (to get a clean stack trace in case there is one being generated), then kill the
+            # remaining processes if needed.
+            sleep 60
```
Comment (Kipok) [inline test/core/execution/artifacts/group_resource_req_slurm.sh:66]: Is it possible to make this and below sleep time customizable? Again, mostly necessary to allow quick debugging and iteration. Currently with all sleep and wait times it's somehow taking longer than 5 minutes for the job to get cancelled after the main task is over.

**Comment by hemildesai** [comment]
@Kipok all comments should be addressed in https://github.com/NVIDIA/NeMo-Run/pull/24/commits/da4c4a4c19c746301298e133993fa1bb0206c6a7

File: src/nemo_run/core/execution/slurm.py (hunk 1 above)
Comment (hemildesai) [reply src/nemo_run/core/execution/slurm.py]: Moved custom srun args to the end

File: test/core/execution/artifacts/group_resource_req_slurm.sh (hunk 2 above)
Comment (hemildesai) [reply test/core/execution/artifacts/group_resource_req_slurm.sh:64]: Done.

**Comment by Kipok** [comment]
Thanks, looks good! The only optional, but nice to have if it's easy feature is to be able to customize logs across parallel jobs (so be able to use different job_paths_cls inside task group). But if it's not that simple to support, we can skip it for now

**Review by marcromeyn — APPROVED** [review]

**Pull request merged**